### PR TITLE
Shipment database structure

### LINKED ISSUE #6351: Asset Management - Asset Shipping support
### NOTE:  For BHIMA, a "Shipment" is defined a **_SINGLE CONTAINER_**.

Therefore several "shipments" may be necessary to send a set of assets to a destination corresponding to a specific requisition.

Note that it is not clear yet if any changes will be needed for requisitions.  It seems likely changes will be necessary.  As you implement the shipping features, please make notes in this Issue (or a new issue) about this.

### Database Changes
- Create a new Shipper (transport) table
  - SQL/functions like Supplier table
- Create new ‘shipment’ table
  - uuid
  - name
  - description
  - shipper_id
  - requistion_id
  - notes
  - current location (depot)
  - destination (depot)
  - anticipated_delivery_date (date)
  - date_sent (ie, date it left a depot for transport elsewhere)
  - date_delivered (ie date it arrived at a depot)
  - receiver at final depot (simple text field to include in printed shipment label)
  - Status (defined in separate table: empty, partial, complete, in-transit, at-depot, delivered, lost)
  - created_at (date)
- Create new ‘shipment_item’ table
  - shipment_uuid
  - asset_uuid (this is the **uuid** in lot table)
  - date_packed
  - notes?
 
### Interface Changes
- Add a new entry “Asset Shipping” under the top-level Assets Management menu group
- Under the “Asset Shipping” menu item add these pages:
    - Shipping Registry
      - Shows grid of shipments, include total value of shipment in a summary.  User can select the display currency.
      - Big button on top:  [New Shipment] or [Add]
        - Brings up a shipment create modal
      - Menu:
        - Search options
      - shipment action menu
        - Edit
        - Scan barcode to change location, status, etc.
        - Print shipping label
        - Print barcode

### PR DESCRIPTION
**Pull request by mbayopanda**
This PR:
- Adds tables for shipment
- Write into shipment during a stock exit
- Update shipment status during stock entry
- Fix stock entry receipt
- Adds shipment in transit (or partial) in the inventory registry

Closes https://github.com/IMA-WorldHealth/bhima/issues/6351

image

How to test:
- Do stock exit and check inventory registry
- Do stock entry and check inventory registry
- Check the stock entry receipt

UPDATE 2022/02/18

We have added :
- Shipment menu
- Shipment registry
- Shipment creation page

The shipment menu in the navigation tree:
image

The shipment registry
image

Actions for a shipment:
image

Creation page for shipment:
image

An overview receipt for a shipment:
image

A report for shipments registry
image

ADDITIONS:
- We have added the table `location` which contains all `depots` by default and we can add other location inside without touching the depot table. This table need a CRUD module

HOW TO TEST:
- Build the database
- Create multiple shipments
- Explore the shipment registry

NOTE: DO NOT MERGE, THIS IS A TRIAL FEEL FREE TO GIVE SUGGESTIONS

### CODE REVIEW

File: client/src/modules/stock/inventories/templates/depot.cell.html [hunk 1]
```diff
@@ -0,0 +1,9 @@
+<div ng-if="!row.groupHeader" class="ui-grid-cell-contents">
+  <span ng-if="!grid.appScope.isInTransitOrPartial(row.entity.shipment_status)">{{row.entity.depot_text}}</span>
+  <span ng-if="grid.appScope.isInTransitOrPartial(row.entity.shipment_status)">
```
Comment (jniles) [inline client/src/modules/stock/inventories/templates/depot.cell.html]: Instead of making a function call here, can you just label each row as it is downloaded in the `read()` function?

So:

```js

glb.inventories.forEach(article => {
  article._isInTransitOrPartial = isInTransitOrPartial(article.shipment_status);
});
```

That would make it cleaner and a bit quicker.

File: server/controllers/stock/reports/common.js [hunk 2]
```diff
@@ -85,7 +85,7 @@ async function getDepotMovement(documentUuid, enterprise, isExit) {
 
   const sql = `
     SELECT
-      i.code, i.text, iu.text AS unit, BUID(m.document_uuid) AS document_uuid,
+      BUID(l.uuid) AS uuid, i.code, i.text, iu.text AS unit, BUID(m.document_uuid) AS document_uuid,
```
Comment (jniles) [inline server/controllers/stock/reports/common.js]: Can you label this `lot_uuid`?  Otherwise, it gets confused with the `stock_movement`.`uuid`, which what I would expect this function to return.

File: server/controllers/stock/shipment.js [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+      shipment_uuid : SHIPMENT_UUID,
+      lot_uuid : db.bid(lot.uuid),
+      date_sent : document.date,
+      quantity_sent : lot.quantity,
+    };
+    transaction.addQuery('INSERT INTO shipment_item SET ?', [shipmentItem]);
+  });
+};
+
+exports.writeStockEntryShipment = async (
```
Comment (jniles) [inline server/controllers/stock/shipment.js]: Do you need this to be `async`?

File: server/controllers/stock/shipment.js [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+  document,
+  parameters,
+  transaction,
+) => {
+  // update shipment items
+  parameters.lots.forEach(lot => {
+    const updateShipmentItem = `
```
Comment (jniles) [inline server/controllers/stock/shipment.js]: This `updateShipmentItem` doesn't change for  each lot.  You can move it outside the loop.


```js
const updateShipmentItem = `
        UPDATE shipment_item shi
        JOIN shipment sh ON sh.uuid = shi.shipment_uuid 
        SET
          shi.quantity_delivered = shi.quantity_delivered + ?,
          shi.date_delivered = ?,
          sh.date_delivered = ?
        WHERE sh.status_id IN (?) AND sh.document_uuid = ? AND shi.lot_uuid = ?
`;

parameters.lots.forEach(lot => {
  transaction.addQuery(updateShipmentItem, [ 
    lot.quantity || 0,
    document.date,
    document.date,
    SHIPMENT_IN_TRANSIT_OR_PARTIAL,
    db.bid(document.uuid),
    db.bid(lot.uuid),
  ]);
});
```

File: server/models/migrations/next/migrate.sql [hunk 5]
```diff
@@ -47,3 +47,87 @@ UPDATE unit SET `name` = 'Analysis of Cashbox', `key` = 'REPORT.ANALYSIS_AUX_CAS
...
+  `destination_depot_uuid`    BINARY(16),
+  `anticipated_delivery_date` DATE,
+  `date_sent`                 DATE,
+  `date_delivered`            DATE,
+  `receiver`                  VARCHAR(100),
+  `status_id`                 SMALLINT(5) UNSIGNED NOT NULL,
+  `created_by`                SMALLINT(5) UNSIGNED NOT NULL,
+  `created_at`                TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,
+  `updated_by`                SMALLINT(5) UNSIGNED NULL,
+  `updated_at`                TIMESTAMP NULL,
```
Comment (jniles) [inline server/models/migrations/next/migrate.sql]: This should have ON UPDATE CURRENT_TIMESTAMP or something similar.

File: server/models/migrations/next/migrate.sql [hunk 6]
```diff
@@ -47,3 +47,87 @@ UPDATE unit SET `name` = 'Analysis of Cashbox', `key` = 'REPORT.ANALYSIS_AUX_CAS
...
+  PRIMARY KEY (`uuid`),
+  CONSTRAINT `shipment__project` FOREIGN KEY (`project_id`) REFERENCES `project` (`id`),
+  CONSTRAINT `shipment__shipper` FOREIGN KEY (`shipper_id`) REFERENCES `shipper` (`id`),
+  CONSTRAINT `shipment__requisition` FOREIGN KEY (`requisition_uuid`) REFERENCES `requisition` (`uuid`),
+  CONSTRAINT `shipment__origin_depot` FOREIGN KEY (`origin_depot_uuid`) REFERENCES `depot` (`uuid`),
+  CONSTRAINT `shipment__current_depot` FOREIGN KEY (`current_depot_uuid`) REFERENCES `depot` (`uuid`),
+  CONSTRAINT `shipment__destination_depot` FOREIGN KEY (`destination_depot_uuid`) REFERENCES `depot` (`uuid`),
+  CONSTRAINT `shipment__status` FOREIGN KEY (`status_id`) REFERENCES `shipment_status` (`id`),
+  CONSTRAINT `shipment__created_by` FOREIGN KEY (`created_by`) REFERENCES `user` (`id`),
+  CONSTRAINT `shipment__updated_by` FOREIGN KEY (`updated_by`) REFERENCES `user` (`id`)
```
Comment (jniles) [inline server/models/migrations/next/migrate.sql:155]: There will be many updates - instead of haivng an `updated_by` column, lets add a separate log for tracking the shipment.

File: server/models/migrations/next/migrate.sql [hunk 7]
```diff
@@ -47,3 +47,87 @@ UPDATE unit SET `name` = 'Analysis of Cashbox', `key` = 'REPORT.ANALYSIS_AUX_CAS
...
+) ENGINE=InnoDB DEFAULT CHARACTER SET = utf8mb4 DEFAULT COLLATE = utf8mb4_unicode_ci;
+
+DROP TABLE IF EXISTS `shipment_item`;
+CREATE TABLE `shipment_item` (
+  `uuid`               BINARY(16) NOT NULL,
+  `shipment_uuid`      BINARY(16) NOT NULL,
+  `lot_uuid`           BINARY(16) NOT NULL,
+  `date_packed`        DATE,
+  `date_sent`          DATE,
```
Comment (jniles) [inline server/models/migrations/next/migrate.sql]: Why do we need `date_sent` and `date_delivered`?  If it is all one shipment, they will all be sent and delivered together, right?

**Review by jniles — COMMENTED** [review]
Just a few comments initially.  I haven't run the code.

Can you include some images of what this looks like?

File: server/models/migrations/next/migrate.sql (hunk 7 above)
Comment (jmcameron) [reply server/models/migrations/next/migrate.sql]: Right.  These should be in the shipment itself, not its items.

File: server/models/triggers.sql [hunk 8]
```diff
@@ -114,9 +114,21 @@ FOR EACH ROW BEGIN
   ON DUPLICATE KEY UPDATE uuid = NEW.document_uuid;
 END$$
 
--- Stock Requisition Triggers
-CALL add_column_if_missing("stock_requisition", "project_id", "SMALLINT(5) UNSIGNED NOT NULL AFTER `user_id`");
+-- Shipment Triggers
+-- the shipment reference is incremented based on the shipment uuid.
+CREATE TRIGGER shipment_reference BEFORE INSERT ON shipment
+FOR EACH ROW
+  SET NEW.reference = (SELECT IF(NEW.reference, NEW.reference, IFNULL(MAX(sH.reference) + 1, 1)) FROM shipment sh WHERE sh.uuid <> NEW.uuid);$$
```
Comment (jmcameron) [inline server/models/triggers.sql]: TYPO!  Note the 'sH.reference' should be 'sh.reference'.  I had to fix this before I could run the tests.

File: client/src/modules/stock/inventories/registry.js [hunk 9]
```diff
@@ -188,6 +191,11 @@ function StockInventoriesController(
   vm.openColumnConfigModal = openColumnConfigModal;
   vm.clearGridState = clearGridState;
 
+  // shipment in transit
+  vm.isInTransitOrPartial = status => {
+    return bhConstants.shipmentStatus.IN_TRANSIT === status || bhConstants.shipmentStatus.PARTIAL === status;
```
Comment (jmcameron) [inline client/src/modules/stock/inventories/registry.js]: A very minor stylistic thing:  Usually, the item being checked is first and the constant it is being checked against is second.  Eg:

   status === bhConstants.shipmentStatus.IN_TRANSIT

I think this reads more naturally and easier for someone reading the code to understand.

**Review by jmcameron — COMMENTED** [review]
This looks like a good first step to me.  Good job!

After fixing the typo in triggers.sql and rebuilding the database, I was able to run this and it seems to work okay.

The test directions should say to check "Articles in Stock", not the inventory registry.

I see that there are several empty files.  I understand that this is a work in progress, but it would be nice to have a single comment in the files indicating what the purpose of each file is.

Should the shipping status be indicated in the stock movement registry?

I'm looking forward to seeing a Shipping registry!

**Comment by mbayopanda** [comment]
> This looks like a good first step to me. Good job!
> 
> After fixing the typo in triggers.sql and rebuilding the database, I was able to run this and it seems to work okay.
> 
> The test directions should say to check "Articles in Stock", not the inventory registry.
> 
> I see that there are several empty files. I understand that this is a work in progress, but it would be nice to have a single comment in the files indicating what the purpose of each file is.
> 
> Should the shipping status be indicated in the stock movement registry?
> 
> I'm looking forward to seeing a Shipping registry!

Thank you @jmcameron for the review, there are several things which are not yet implemented and the database structure can change a bit over the time.

File: client/src/i18n/en/shipment.json [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+    "DELETE": "Delete",
+    "REFERENCE": "Reference",
+    "STOCK_REFERENCE": "Stock Reference",
+    "PACKING": "Packing",
+    "PACKING_LIST": "Packing List",
+    "NEW_SHIPMENT": "New Shipment",
+    "SHIPMENT_REGISTRY": "Shipment Registry",
+    "UPDATE_LOCATION": "Update Location",
+    "UPDATE_LOCATION_MESSAGE": "You are about to change the location of this shipment, this information will be saved in the history of this shipment",
+    "UPDATE_LOCATION_WARNING": "You cannot change the location because only shipments with an IN-TRANSIT status can benefit from the change of location",
```
Comment (jmcameron) [inline client/src/i18n/en/shipment.json]: Suggest rewording.  In "only shipments with an IN-TRANSIT status can benefit from a change in location" change "benefit" to "allow".

**Review by jmcameron — COMMENTED** [review]
I tested this using bhima_test and performed tests based on the suggestions.   
I did not do a careful code walkthrough.

Since shipments are related to stock in general and not just
assets, I think the Shipments menu should not be under "Asset
Management".  Perhaps it should just be at the top level?

When we have a stock exit and we look at the items involved in the
Articles in Stock registry, it shows good information about the
shipment in the Depot column.  However, it would be helpful to add
a Tooltip with the same information so when you hover over it with
your mouse, you will see the entire message easily.

When we create a shipment by doing a stock exit, it sets an
anticipated delivery date automatically (using the same date).
However, it seems like you may want to change this in some cases.
If I go to the shipment registry, "Edit Shipment" is not an option
in the action menu.  How can that date be changed?

The Shipment Overview is nice, but I don't see the anticipated
delivery date on it.

I tested using the directions.  The stock exit worked.  And when
I look at the Articles in Stock registry, I could see the
shipment. That works well. However, when I tried the stock entry I
got this error (it looks like no value is provided for
current_depot_uuid): 

   "You have an error in your SQL syntax; check the manual that
   corresponds to your MySQL server version for the right syntax to
   use near:

     UPDATE shipment SET status_id = 6, current_depot_uuid WHERE
     document_uuid = X'd4bb1452e4fa4742a281814140246877';

**Comment by jmcameron** [comment]
I took another look after the fix in https://github.com/IMA-WorldHealth/bhima/pull/6394/commits/769f64defe19c114248b3e303a5c9fd27dfe1cca.  I was able to do the stock entry and it all seemed to work.

In the shipments registry, I suggest adding a filter that will hide any completed shipments.

I think the shipment overview document should clearly and boldly say the current status of the order somewhere.  I realize that is in the graphic, but adding a separate line that says something like this would make it very clear:

   Status:  In-transit

perhaps this could be right under the date but above the shipment status graphic?

I'm not sure if this is easy or make sense, but would it be possible to put the date below each item in the shipment status graphic?

I liked the graphic for the shipment status.

**Comment by mbayopanda** [comment]
Thank you @jmcameron for the review, I am working on that. Thank you

**Comment by mbayopanda** [comment]
We have two ways of handling shipment:
- With the stock exit to depot: the traditional way which is working 100%
- With the create shipment module: a tool which allow to add much more information in a shipment, and allow to edit the shipment.

The remaining part is to handle the stock exit of shipment created with the create-shipment module

**Comment by mbayopanda** [comment]
We have added a new option in the "Find Depot Modal" for shipment: 
image

The remaining part is to find a way to lock or disable the grid once data are loaded for shipment.

**Comment by mbayopanda** [comment]
bors try

**Comment by jniles** [comment]
@mbayopanda this still have "do not merge" so I've been ignoring it.  Let me know when you are ready for a review 👍

**Comment by mbayopanda** [comment]
> @mbayopanda this still have "do not merge" so I've been ignoring it. Let me know when you are ready for a review 👍

Let me rebase with the last commit from your PR.

**Comment by mbayopanda** [comment]
@jniles @jmcameron when you'll be available you can review this PR

File: client/src/modules/shipment/create-shipment.js [hunk 11]
```diff
@@ -0,0 +1,321 @@
...
+
+    const isValidForSubmission = vm.stockForm.validate();
+
+    if (!isValidForSubmission) { return null; }
+
+    vm.$loading = true;
+    vm.shipment = cleanShipment(vm.shipment);
+    vm.shipment.lots = vm.stockForm.store.data.map(cleanShipmentItem);
+
+    const promise = (vm.isCreateState)
```
Comment (jmcameron) [inline client/src/modules/shipment/create-shipment.js]: Note: The Javascript for the ternary (conditional) expression here is correct, but generally the condition part is only in parenthesis if an expression, not a singly truthy variable.

**Review by jmcameron — COMMENTED** [review]
I glanced through the code and it seems okay.

**I tested this with bhima_test:**
- Created a shipment (I had one problem, see below)
- Marked the shipment as ready to ship (good)
- Did stock exit based on the shipment (this worked fine)
- Added shipment transit log item (good)
- Looked at Stock > Articles in Stock and saw the items in transit (good)
- Could not figure out how to do the stock entry from a shipment

**Problems**
- It was not clear to from the PR list of suggested changes, etc, which ones have been addressed.   I suggest you manually go through all the items and mark them as "Resolved" if you have fixed them, fix any unaddressed items, or indicate the the item will not be fixed (and why). 
- While I was creating a shipment, it allowed me to add expired items.  This needs to be fixed.
- Ran into an unrelated bug:
  - Go to Stock > Entry
  - Open up the browser Javascript error console
  - Click on Purchase.  Then click Cancel.
  - Note the client-side error: could not find 'currency_id' (in findPurchase, entry.js line 370)
- I like the shipment progress timeline.  Very nice!  But I may be missing something, 
  but it seems like we need a way to manually note that a shipment has been received 
  on the receiving end and is ready for stock entry.

File: client/src/modules/stock/inventories/templates/depot.cell.html (hunk 1 above)
Comment (mbayopanda) [reply client/src/modules/stock/inventories/templates/depot.cell.html]: RESOLVED 👍🏽

File: server/controllers/stock/reports/common.js (hunk 2 above)
Comment (mbayopanda) [reply server/controllers/stock/reports/common.js]: RESOLVED 👍🏽

File: server/controllers/stock/shipment.js (hunk 3 above)
Comment (mbayopanda) [reply server/controllers/stock/shipment.js]: RESOLVED 👍🏽

File: server/controllers/stock/shipment.js (hunk 4 above)
Comment (mbayopanda) [reply server/controllers/stock/shipment.js]: RESOLVED 👍🏽

File: server/models/migrations/next/migrate.sql (hunk 5 above)
Comment (mbayopanda) [reply server/models/migrations/next/migrate.sql]: RESOLVED 👍🏽

File: server/models/migrations/next/migrate.sql (hunk 6 above)
Comment (mbayopanda) [reply server/models/migrations/next/migrate.sql:155]: RESOLVED 👍🏽

File: server/models/migrations/next/migrate.sql (hunk 7 above)
Comment (mbayopanda) [reply server/models/migrations/next/migrate.sql]: RESOLVED 👍🏽

File: server/models/triggers.sql (hunk 8 above)
Comment (mbayopanda) [reply server/models/triggers.sql]: RESOLVED 👍🏽

File: client/src/modules/stock/inventories/registry.js (hunk 9 above)
Comment (mbayopanda) [reply client/src/modules/stock/inventories/registry.js]: RESOLVED 👍🏽

**Comment by mbayopanda** [comment]
Thank you @jmcameron for the review

**Comment by mbayopanda** [comment]
@jmcameron changes landed

**Review by jmcameron — COMMENTED** [review]
I tested this again.  Here are problems I encountered
- While creating a shipment, it allows me to add an expired lot.   I suggest creating a new issue for this so we can fix it later.
- We need to make it easier for the person receiving a shipment for a depot to figure how to select a document corresponding to a shipment.

**Review by jmcameron — APPROVED** [review]
Looks good.   I have a few suggestions, but we can fix them later.

Please do a 'github rebase' on it and push a new version with squashed commits.  Then go ahead and release it.  Thanks!

**Comment by jniles** [comment]
bors ping
